### PR TITLE
Avoid reporting ref safety errors for discard assignments

### LINKED ISSUE #65522: Ref safety errors reported for deconstruction assignment with discard
Compile with VS 17.4.0 / csc 4.4.0-3.22518.13 (7856a68c):
```csharp
using System;
class Program
{
    static void F1()
    {
        Span<int> s1 = stackalloc int[10];
        s1.Deconstruct(out s1, out _); // ok
    }
    static void F2()
    {
        Span<int> s2 = stackalloc int[10];
        (s2, _) = s2; // errors
    }
}
static class Extensions
{
    public static void Deconstruct(this Span<int> self, out Span<int> x, out Span<int> y)
    {
        throw null;
    }
}
```

**Expected Behavior**:
No errors

**Actual Behavior**:
(12,9): error CS8352: Cannot use variable '(s2, _) = s2' in this context because it may expose referenced variables outside of their declaration scope
(12,19): error CS8350: This combination of arguments to 'Extensions.Deconstruct(Span<int>, out Span<int>, out Span<int>)' is disallowed because it may expose variables referenced by parameter 'self' outside of their declaration scope

### LINKED ISSUE #65651: Unexpected ref safety error reported with `out _` with C#7.2 ref safety rules
Compile with .NET 6 and `-langversion:10`:
```csharp
using System;

class Program
{
    static Span<int> Test1()
    {
        var s = ReturnsSpan(out _);
        return s; // error CS8352: Cannot use variable 's' in this context ...
    }

    static Span<int> Test2()
    {
        var s = ReturnsSpan(out var _);
        return s;
    }

    static Span<int> Test3()
    {
        var s = ReturnsSpan(out var unused);
        return s;
    }

    static ref Span<int> ReturnsSpan(out Span<int> x)
    {
        x = default;
        return ref x;
    }
}
```

**Expected Behavior**:
No errors

**Actual Behavior**:
(8,16,8,17): error CS8352: Cannot use variable 's' in this context because it may expose referenced variables outside of their declaration scope

### PR DESCRIPTION
**Pull request by cston**
Fixes #65522
Fixes #65651

Also fixes an additional issue where the compiler was reporting a ref safety error for deconstruction into an explicitly typed discard - see `RefEscapingTests.Discard_03`:
```csharp
(var x3, R _) = F(); // error CS8350: ... may expose variables referenced by parameter 'y' ...

ref struct R 
{
    public void Deconstruct(out R x, out R y) => throw null;
}
```

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs [hunk 1]
```diff
@@ -182,7 +182,7 @@ internal override BoundStatement BindForEachDeconstruction(BindingDiagnosticBag
             ExpressionSyntax variables = ((ForEachVariableStatementSyntax)_syntax).Variable;
 
             // Tracking narrowest safe-to-escape scope by default, the proper val escape will be set when doing full binding of the foreach statement
-            var valuePlaceholder = new BoundDeconstructValuePlaceholder(_syntax.Expression, variableSymbol: null, this.LocalScopeDepth, inferredType.Type ?? CreateErrorType("var"));
+            var valuePlaceholder = new BoundDeconstructValuePlaceholder(_syntax.Expression, variableSymbol: null, this.LocalScopeDepth, isDiscardExpression: false, inferredType.Type ?? CreateErrorType("var"));
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:185]: >false [](http://example.com/codeflow?start=157&length=5)

I didn't follow this. It should be possible to have a discard in a deconstruction-foreach

File: src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs [hunk 2]
```diff
@@ -14339,6 +14321,86 @@ static void verify(CSharpCompilation comp)
...
+        public void LocalScope_01_Foreach_Deconstruction_02(LanguageVersion languageVersion)
+        {
+            var source =
+@"#pragma warning disable 219
+
+class Program
+{
+    static void F(ref R r)
+    {
+        foreach ((R r0, _) in new Enumerable1()) break;
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:14336]: Consider adding a non-deconstruction variant: `foreach (var _ in ...)`

Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RefFieldTests.cs:14336]: Also, I liked how for deconstruction tests you added the equivalent lowered version (calling `Deconstruct`). Consider doing the same for foreach-deconstruction scenarios, adding equivalent lowered code.

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 4)

File: src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs (hunk 1 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:185]: I believe this placeholder is for the expression rather than iteration variables.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 6)

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**